### PR TITLE
Add Py2x flag

### PR DESCRIPTION
**Pull request by nanjekyejoannah**
This PR adds the Py2x flag:

```
Python 3.12.0a0 (heads/migration-dirty:5842fee697, May 23 2022, 00:25:04) [Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.flags
sys.flags(debug=0, inspect=0, interactive=0, optimize=0, dont_write_bytecode=0, no_user_site=0, no_site=0, ignore_environment=0, verbose=0, bytes_warning=0, py2x_warning=1, quiet=0, hash_randomization=1, isolated=0, dev_mode=False, utf8_mode=0, warn_default_encoding=0)
>>> 
```

### CODE REVIEW

File: Doc/c-api/exceptions.rst [hunk 1]
```diff
@@ -1091,6 +1092,8 @@ the variables:
 +------------------------------------------+---------------------------------+----------+
 | :c:data:`PyExc_BytesWarning`             | :exc:`BytesWarning`             |          |
 +------------------------------------------+---------------------------------+----------+
+| :c:data:`PyExc_Py2xWarning`             | :exc:`Py2xWarning`             |          |
```
Comment (ltratt) [inline Doc/c-api/exceptions.rst]: I think the alignment of this row has gone a bit off and a bit more whitespace needs to be added?

File: Lib/subprocess.py [hunk 2]
```diff
@@ -298,6 +298,7 @@ def _args_from_interpreter_flags():
         'verbose': 'v',
         'bytes_warning': 'b',
         'quiet': 'q',
+        #'py2x_warning': '2',
```
Comment (ltratt) [inline Lib/subprocess.py]: Intentionally commented out?

Comment (nanjekyejoannah) [reply Lib/subprocess.py]: Thanks, it has no effect but I may have to just remove it rather than comment.

Comment (ltratt) [reply Lib/subprocess.py]: Unless there's a good reason for commented code, I prefer to delete it (commented code is a sort-of phobia of mine!).

File: Doc/library/exceptions.rst [hunk 3]
```diff
@@ -855,6 +855,10 @@ The following exceptions are used as warning categories; see the
 
    Base class for warnings related to :class:`bytes` and :class:`bytearray`.
 
+.. exception:: Py2xWarning
+
+   Base class for warnings related to 3.x compatibility.
```
Comment (vext01) [inline Doc/library/exceptions.rst:860]: Should it be "related to 2.x compatibility"?

File: Lib/test/test_cmd_line.py [hunk 4]
```diff
@@ -684,7 +684,7 @@ def test_xdev(self):
         # Warnings
         code = ("import warnings; "
                 "print(' '.join('%s::%s' % (f[0], f[2].__name__) "
-                                "for f in warnings.filters))")
+                                "for f in warnings.filters))")
```
Comment (vext01) [inline Lib/test/test_cmd_line.py]: Is this change intentional?

**Comment by vext01** [comment]
Just a couple of comments from me.

I'm surprised how much boilerplate is required, but it looks correct to me.

File: Doc/library/exceptions.rst (hunk 3 above)
Comment (nanjekyejoannah) [reply Doc/library/exceptions.rst:860]: Seems that we are warning when one write 2.x code, to signal for 3.x compatibility.

File: Lib/test/test_cmd_line.py (hunk 4 above)
Comment (nanjekyejoannah) [reply Lib/test/test_cmd_line.py]: Not, I will use `make patchcheck` to remove these spaces.

**Comment by nanjekyejoannah** [comment]
Great, thanks, If its in good shape, I will just add some tests for `_embed` and the `sys` module, then I will use the flag for the `cmp` porting (In a second PR).

File: Doc/library/exceptions.rst (hunk 3 above)
Comment (vext01) [reply Doc/library/exceptions.rst:860]: So it's correct? If so, OK :)

**Comment by ltratt** [comment]
@nanjekyejoannah Works for me!

**Comment by nanjekyejoannah** [comment]
I have addressed the changes and fixed the flag bug. Multiprocessing tests were exhibiting random behaviours, so I would appreciate, a review on that.

**Comment by ltratt** [comment]
This is OK with me. @vext01 ?

File: Lib/test/_test_multiprocessing.py [hunk 5]
```diff
@@ -3965,18 +3965,18 @@ def test_shared_memory_recreate(self):
...
 
             mock_make_filename.side_effect = names
-            shm1 = shared_memory.SharedMemory(create=True, size=1)
+            shm1 = shared_memory.SharedMemory(create=True, size=512)
             self.addCleanup(shm1.unlink)
             self.assertEqual(shm1._name, names[0])
 
             mock_make_filename.side_effect = names
-            shm2 = shared_memory.SharedMemory(create=True, size=1)
+            shm2 = shared_memory.SharedMemory(create=True, size=512)
```
Comment (vext01) [inline Lib/test/_test_multiprocessing.py]: we've changed some filenames and parameters here. Is that needed?

Comment (vext01) [reply Lib/test/_test_multiprocessing.py]: To elaborate, I don't see why adding a py2x flag would require these path and size changes. Maybe I missed something.

Comment (nanjekyejoannah) [reply Lib/test/_test_multiprocessing.py]: I increased from 1 to 512, this test was failing so much for recreation cases. That is why I need some verification.

Comment (vext01) [reply Lib/test/_test_multiprocessing.py]: Was it failing before adding the flag?

Comment (nanjekyejoannah) [reply Lib/test/_test_multiprocessing.py]: Not, but sometimes it gets skipped, so its random, it can run or not run, if you run the whole test suite. Anyway, I can revert it, no big deal.

Comment (vext01) [reply Lib/test/_test_multiprocessing.py]: Where did you get the idea to rename the files and up the size? Is that something you saw in the wild?

Comment (nanjekyejoannah) [reply Lib/test/_test_multiprocessing.py]: Yes, the recreate test, would fail at times, the renaming, might have come with other changes that fixed it though. A particular conditional I removed in a config setup.

**Comment by ltratt** [comment]
Should I do a bors try?

**Comment by nanjekyejoannah** [comment]
Yes

**Comment by ltratt** [comment]
bors try

**Comment by ltratt** [comment]
I think we're ready to squash?

**Comment by nanjekyejoannah** [comment]
Sure, let me do now

**Comment by ltratt** [comment]
Could we add a bit of motivation to the commit message, even if only a couple of sentences? Please feel free to force push an update.

**Comment by nanjekyejoannah** [comment]
Does this "Add a py2x_warning flag to issue a warning related to 3.x compatibility" look fine?

**Comment by ltratt** [comment]
I am a fan of motivation (i.e. "why" rather than "what" since the latter is implicitly contained in the diff but the former is often hard to divine), so maybe adding a bit more would be good? e.g. "This is the equivalent of the `-2` switch in python2. As with that switch, the aim is to ease Python 2->3 migration, though in this case in the opposite direction to the `-2` switch." I'm sure you can do better than that, but hopefully it's an idea :)

**Comment by nanjekyejoannah** [comment]
I made some edits to your wording in addition to the former "what" message.

**Comment by ltratt** [comment]
We're almost there :)

We just need to add a double newline after "Add a -2 flag to issue a warning related to 3.x compatibility." otherwise git log will render everything terribly. [Commit messages are "Short description\n\nLong description".] Please force push that and then I'll merge :)

**Comment by nanjekyejoannah** [comment]
Ok, how is this?

**Comment by ltratt** [comment]
Perfect!

bors r+
